src/masked-input/masked-input.tsx

```diff
@@ -0,0 +1,154 @@
+import React, { useEffect, useState } from "react";
+import { InputGroup } from "../input-group/input-group";
+import { EyeIcon } from "@lifesg/react-icons/eye";
+import { EyeSlashIcon } from "@lifesg/react-icons/eye-slash";
+import { IconContainer } from "./masked-input.style";
+import { MaskedInputProps } from "./types";
+import { isEmpty } from "lodash";
+
+const Component = <T, V>(
+    {
+        value,
+        readOnly,
+        "data-testid": dataTestId,
+        maskRange,
+        unmaskRange,
+        maskRegex,
+        maskTransformer,
+        onChange,
+        iconMask = <EyeSlashIcon />,
+        iconUnmask = <EyeIcon />,
+        iconActiveColor: maskIconActiveColor,
+        iconInactiveColor: maskIconInactiveColor,
+        maskChar = "•",
+        onMask,
+        onUnmask,
+        error,
+        ...otherProps
+    }: MaskedInputProps<T, V>,
+    ref: React.Ref<HTMLInputElement>
+) => {
+    const isEmptyReadOnlyState = readOnly && isEmpty(value);
+    const [isMasked, setIsMasked] = useState(true);
+    const [updatedValue, setUpdatedValue] = useState(
+        isEmptyReadOnlyState ? "-" : value || ""
+    );
+
+    useEffect(() => {
+        if (isMasked) {
+            onMask && onMask();
+        } else {
+            onUnmask && onUnmask();
+        }
+    }, [isMasked]);
+
+    // =============================================================================
+    // EVENT HANDLERS
+    // =============================================================================
+
+    const handleFocus = () => !readOnly && setIsMasked(false);
+    const handleBlur = () => !readOnly && setIsMasked(true);
```

Comment:
we can propagate the focus and blur events by calling `onFocus` and `onBlur` props here as well